core/src/main/java/com/linecorp/armeria/common/stream/StreamWriterOutputStream.java

```diff
@@ -0,0 +1,104 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpData;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+class StreamWriterOutputStream<T> extends OutputStream {
+
+    private final StreamWriter<T> writer;
+    private final Function<? super HttpData, ? extends T> httpDataConverter;
+    private final ByteBuf buffer = Unpooled.buffer();
+    private final int maxBufferSize;
+    private boolean closed;
+
+    StreamWriterOutputStream(StreamWriter<T> writer,
+                             Function<? super HttpData, ? extends T> httpDataConverter,
+                             int maxBufferSize) {
+        requireNonNull(writer, "writer");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        checkArgument(maxBufferSize > 0, "maxBufferSize should be positive");
+        this.writer = writer;
+        this.httpDataConverter = httpDataConverter;
+        this.maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        ensureOpen();
+        ensureBufferWritable();
+        buffer.writeByte(b);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        ensureOpen();
+        ensureBufferWritable();
+        buffer.writeBytes(b, off, len);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        ensureOpen();
+        drain();
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        flush();
+        writer.close();
+        closed = true;
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+        if (!writer.isOpen()) {
+            closed = true;
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void ensureBufferWritable() {
+        if (buffer.readableBytes() >= maxBufferSize) {
+            drain();
+        }
+    }
+
+    private void drain() {
+        if (!buffer.isReadable() || !writer.isOpen()) {
+            return;
+        }
+        final byte[] data = new byte[buffer.readableBytes()];
+        buffer.readBytes(data);
+        writer.write(httpDataConverter.apply(HttpData.copyOf(data)));
+    }
```

Comment:
> How about using an unpooled heap ByteBuf so it is always backed by a byte array? Then, we could use ByteBuf.array() to access the underlying array and copy it into a new HttpData using HttpData.copyOf(byte[],...).

hmm.. with ByteBuf.array(), the underlying array is not I want 🤔
```java
// what I wanted for data: { 1 };
// But was { 1, 0, 0, 0, 0, ... , 0 }
final byte[] data = buffer.readBytes(Unpooled.buffer()).array(); 
writer.write(httpDataConverter.apply(HttpData.copyOf(data)));
```

Reply:
You may want to refer to https://github.com/line/armeria/blob/5b384fbe27e7e6f9225d6db91cbb684d09dfbb5e/core/src/main/java/com/linecorp/armeria/common/ByteBufHttpData.java#L60

Reply:
Actually I'm newbie on `Netty#ByteBuf` so I can't fully understand this comment 😅 
Anyway, I implemented `drain()` via `buffer.readtBytes(byte[])`.

If there's better way to readBytes from buffer, then please share to me!

```java
    private void drain() {
        if (!buffer.isReadable() || !writer.isOpen()) {
            return;
        }
        // If there's better way to readBytes from buffer, then please share to me!
        final byte[] data = new byte[buffer.readableBytes()];
        buffer.readBytes(data);
        writer.write(httpDataConverter.apply(HttpData.wrap(data)));
    }
```